lib/rules/layer-name-pattern/index.mjs

```diff
@@ -0,0 +1,108 @@
+import { atRuleRegexes, functionRegexes } from '../../utils/regexes.mjs';
+import { isRegExp, isString } from '../../utils/validateTypes.mjs';
+import { atRuleParamIndex } from '../../utils/nodeFieldIndices.mjs';
+import report from '../../utils/report.mjs';
+import ruleMessages from '../../utils/ruleMessages.mjs';
+import validateOptions from '../../utils/validateOptions.mjs';
+
+const ruleName = 'layer-name-pattern';
+
+const messages = ruleMessages(ruleName, {
+	expected: (name, pattern) => `Expected "${name}" to match pattern "${pattern}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/layer-name-pattern',
+};
+
+/** @type {import('stylelint').CoreRules[ruleName]} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [isRegExp, isString],
+		});
+
+		if (!validOptions) return;
+
+		const pattern = isString(primary) ? new RegExp(primary) : primary;
+
+		root.walkAtRules(atRuleRegexes.layer, (atRule) => {
+			if (!atRule.params) return;
+
+			const layerNames = atRule.params.split(',').map((s) => s.trim());
+			const baseIndex = atRuleParamIndex(atRule);
+			let startIndex = 0;
+
+			layerNames.forEach((layerName) => {
+				const segments = layerName.split('.');
+
+				segments.forEach((segment) => {
+					if (pattern.test(segment)) return;
+
+					const segmentStart = atRule.params.indexOf(segment, startIndex);
+
+					const index = baseIndex + segmentStart;
+					const endIndex = index + segment.length;
+
+					report({
+						message: messages.expected(segment, primary),
+						node: atRule,
+						index,
+						endIndex,
+						ruleName,
+						result,
+					});
+				});
+
+				// Added +1 because the next layer name starts with a comma
+				startIndex = atRule.params.indexOf(layerName) + layerName.length + 1;
+			});
+		});
+
+		root.walkAtRules(atRuleRegexes.importName, (atRule) => {
+			const layerMatch = atRule.params.match(functionRegexes.layer);
```

Comment:
Let's [use](https://stylelint.io/developer-guide/rules#construct-specific-parsers) `postcss-value-parser` to find the `layer` function and get the comma-separated names.